# <font color=black>DATA 557 Homework 2</font>
***
### <font color=black>Question 1
A manufacturing process is run at a temperature of 60 deg C. The manufacturer would like to know if 
increasing the temperature would yield an increase in output. Increasing the temperature would be more 
expensive, so an increase would only be used in future if it increased output. It seems unlikely that 
increasing the temperature would decrease output and, even if it did, there would be no value in having 
that information. An experiment was performed to assess the effect of temperature on the output of a 
manufacturing process. For this experiment, temperatures of 60 or 75 degrees C were randomly assigned 
to process runs. It was desired to gather more information about output at the new temperature so 
temperatures were randomly assigned to process runs at a ratio of 2 to 1 (2 runs at temperature 75 for 
every 1 at temperature 60). The process output was recorded from each run. The variables in the data set 
are: \
run: Run number \
temp: Temperature \
output: Process output</font>

__1.1. Perform the large-sample Z-test to compare mean output for the two temperatures. Give the value of 
the test statistic and the p-value for the test.__

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import statistics as stat
import numpy as np
from scipy import stats

Since the question specifies that there is no value in having information on decreasing output due to an increase in temperatures, I am considering a __One Tailed large-sample Z-test__.
<br>
We can define the null hypothesis as : <br>
<center> H0 : μ75 - μ60 <= 0 <br>
                           H1 : μ75 - μ60 > 0 </center>

In [2]:
df = pd.read_csv("data/temperature_experiment.csv")
temp60 = df.loc[(df["temp"] == 60)]
temp75 = df.loc[(df["temp"] == 75)]
var60 = stat.stdev(temp60['output'])**2
var75 = stat.stdev(temp75['output'])**2
print(len(temp60))
test_statistic = (temp75["output"].mean()-temp60["output"].mean())/(((var75/20)+(var60/10))**0.5) 
p_val = stats.norm.sf(abs(test_statistic)) 
print('\033[1m'+'One Tailed Z Test : '+'\tTest statistic: %f' %(test_statistic))
print('\033[1m'+'\t\t\tp-value : %f' %(p_val))

10
One Tailed Z Test : 	Test statistic: 2.551155
			p-value : 0.005368


__1.2. Do you reject the null hypothesis at a significance level of 0.05?__

In [63]:
z_crit = stats.norm.ppf(q=0.95)
print('\033[1m'+"Z critical : %f" %(z_crit))
print('\033[1m'+"Significance Level : 0.05")
print('\033[1m'+'p-value : %f' %(p_val))

Z critical : 1.644854
Significance Level : 0.05
p-value : 0.005368


Since p-value is less than a 0.05 significance level, we __Reject__ the null hypothesis.
So we can with 0.05 significance level, we can say that with an increase in temperature, there is an increase in output.

__1.3. State the null hypothesis for the test.__`

 μ75 is the mean output at 60 deg C
 μ60 is the mean output at 75 deg C<br>
We can define the null hypothesis as : <br>
<center> H0 : μ75 - μ60 <= 0 <br>
                           H1 : μ75 - μ60 > 0 </center>
This is a one tailed test, we consider on the right sided test as we only focus on whether the ouput increases with an increase in temperature

__1.4. Perform the unequal-variance (Welch) t-test to compare mean output in the two temperature groups. 
Report the test statistic and the p-value for the test.__

In [70]:
welch_df= (((var75/20) + (var60/10))**2)/(((var75**2)/(400*19))+((var60**2)/(100*9)))
p_val = stats.t.sf(abs(test_statistic), df=welch_df)
#test statistic is the same as what we calculated above
print('\033[1m'+'Welch df : %f '%welch_df)
print('\033[1m'+'Welch t Test (One-tailed): '+'\tTest statistic: %f' %(test_statistic))
print('\033[1m'+'\t\t\t\tp-value : %f' %(p_val))

Welch df : 25.633485 
Welch t Test (One-tailed): 	Test statistic: 2.551155
				p-value : 0.008531


__1.5. Perform the equal-variance t-test to compare mean output in the two temperature groups. Report the 
test statistic and the p-value for the test.__

In [73]:
pooled_var=((19*var75)+(9*var60))/28
test_statistic_t = (temp75["output"].mean()-temp60["output"].mean())/(((pooled_var/20)+(pooled_var/10))**0.5) 
p_val = stats.t.sf(abs(test_statistic_t), df=28)
print('\033[1m'+'Equal Variance t-test:\n\tpooled variance : %f \n\ttest statistic : %f \n\tp value : %f'%(pooled_var,test_statistic,p_val))

Equal Variance t-test:
	pooled variance : 597.387571 
	test statistic : 2.551155 
	p value : 0.032238


__1.6. Which of the three tests do you think is most valid for this experiment? Why?__

In [76]:
print("Variance temp 60 : %f \nVariance temp 75 : %f"%(var60,var75))

Variance temp 60 : 91.569333 
Variance temp 75 : 836.985684


Considering the large sample Z test - the sample sizes are relatively small. Hence, the sample is more likely to fit a t distribution over a nornak distribution.
Considering the equal variance t test - we can clearly see that the variances are not equal, hence making this assumption would be wrong.
Considering Welch t test - I believe, this is the ideal test as we are working with small samples with unequal variances.

For these reasons, Welch __t-test__ is the most valid for this experiment

__1.7. Calculate a 95% confidence interval for the difference between mean output using the large-sample 
method.__

In [84]:
se=(var75/20+var60/10)**0.5
A = (temp75["output"].mean()-temp60["output"].mean())-1.96*se
B = (temp75["output"].mean()-temp60["output"].mean())+1.96*se
print('\033[1m'+'Confidence Interval for large sample Z test is : (%f,%f)' %(A,B)) 
#I have reported the two tailed confidence interval in line 

Confidence Interval for large sample Z test is : (4.221947,32.218053)


__1.8. Calculate a 95% confidence interval for the difference between mean output using a method that 
corresponds to the Welch test.__

In [85]:
t_critical=stats.t.ppf(q=1-0.05/2,df=welch_df)
A = (temp75["output"].mean()-temp60["output"].mean())-(t_critical*se)
B = (temp75["output"].mean()-temp60["output"].mean())+t_critical*se
print('\033[1m'+'Confidence Interval is for Welch t-test is : (%f,%f)' %(A,B))
#I have reported the two tailed confidence interval in line with the announcement

Confidence Interval is for Welch t-test is : (3.529466,32.910534)


__1.9. Calculate a 95% confidence interval for the difference between mean output using a method that 
corresponds to the equal-variance t-test.__

In [86]:
se=(pooled_var*((1/20)+(1/10)))**0.5
t_critical=stats.t.ppf(q=1-0.05/2,df=28)
A = (temp75["output"].mean()-temp60["output"].mean())-(t_critical*se)
B = (temp75["output"].mean()-temp60["output"].mean())+t_critical*se
print('\033[1m'+'Confidence Interval for equal variance t-test is : (%f,%f)' %(A,B))
#I have reported the two tailed confidence interval in line with the announcement

Confidence Interval for equal variance t-test is : (-1.170544,37.610544)


__1.10. Apart from any effect on the mean output, do the results of the experiment suggest a disadvantage of 
the higher temperature?__

In [87]:
print("Output at temperature 60 deg C : \n\tMean : %f \n\tVariance : %f \nOuput at temperature 75 deg C : \n\tMean : %f \n\tVariance : %f"%(temp60["output"].mean(),var60,temp75["output"].mean(),var75))

Output at temperature 60 deg C : 
	Mean : 1001.240000 
	Variance : 91.569333 
Ouput at temperature 75 deg C : 
	Mean : 1019.460000 
	Variance : 836.985684


There is no significant change in the mean, but there is a considerable difference in variance when we increase temperature. It can be more likely that the output at higher temperatures in not as consistent or as stable at the lower temperature. So if the manufacturer, values consistency in their output, this could be a potential disadvantage.

***
### <font color=black>Question 2
The data are from an experiment to compare 4 processing methods for manufacturing steel ball bearings. 
The 4 process methods were run for one day and a random sample of 1% of the ball bearings from the day 
was taken from each of the 4 methods. Because the processes produce ball bearings at different rates the 
sample sizes were not the same for the 4 methods. Each sampled ball bearing had its weight measured to 
the nearest 0.1 g and the number of surface defects was counted. The variables in the data set are: \
Sample: sample number \
Method: A, B, C, or D \
Defects: number of defects \
Weight: weight in g</font>

__2.1. The target weight for the ball bearings is 10 g. For each of the 4 methods it is desired to test the null 
hypothesis that the mean weight is equal to 10. What test should be used?__

For this experiment, since we are only considering a 1% sample from each of the processing methods, where number of observations for each method individually are :
A : 74
B : 75
C : 63
D : 52
I believe, the sample sizes are not large enough to conduct a z test.

Since we want to test if mean weight is 10, and any values below or above this value needs to considered, for each of the 4 methods : <br> performing a __Two Tailed One Sample t Test__ is appropriate.

In [88]:
df = pd.read_csv("data/defects.csv")

__2.2. Give the p-values for the tests for each method. Include your R code for this question.__

We can define the null hypothesis for each test as :
H0 : μ = 10 <br>
H1 : μ not equal to 10
    

In [98]:
def test_stat(df, processing_method):
    df_pm = df.loc[(df["Method"] == processing_method)]
    test_statistic = (df_pm["Weight"].mean()-10)/(stat.stdev(df_pm['Weight'])/(len(df_pm)**0.5))
    p_val = stats.t.sf(abs(test_statistic), df=len(df_pm)-1)*2 
    print("For Processing Method %s :\n\tTest Statistic = %f \n\t\033[1mp-value\033[0m = %f"%(processing_method,test_statistic,p_val))

processing_methods=['A','B','C','D']
for i in processing_methods:
    test_stat(df,i)

For Processing Method A :
	Test Statistic = 0.079652 
	p-value = 0.936732
For Processing Method B :
	Test Statistic = -0.495650 
	p-value = 0.621610
For Processing Method C :
	Test Statistic = 1.486649 
	p-value = 0.142095
For Processing Method D :
	Test Statistic = 2.068224 
	p-value = 0.043708


__2.3. Apply a Bonferroni correction to your results from the previous question to account for inflation of 
type I error rate due to multiple testing. How does the Bonferroni correction change your conclusions? In 
particular, do you have evidence to reject the null hypothesis that the mean weight for all 4 methods is 
equal to 10, at significance level 0.05?__

In [100]:
error = 1 - pow(1-0.05,4)
alpha_bonferroni = 0.05/4 #0.0125
error_new = 1 - pow(1-alpha_bonferroni,4)
print('\033[1m'+"Type I error rate before correction : %f \nType I error rate after correction : %f"%(error,error_new))

Type I error rate before correction : 0.185494 
Type I error rate after correction : 0.049070


In __2.2__, with a significance level of 0.05, we see that for processing methods A,B,C the p-value is greater than 0.05. But for D, the p-value(0.0437) is less than 0.05 which is why we have to __reject__ the null hypothesis that the mean for all methods is equal to 10. Also, before the correction due to multiple tests, the Type I error rate is 0.185.

Now, __After Bonferroni Correction__, the new significance level is 0.0125 for each individual test, All the p-values from the processing methods (A,B,C,D) are greater than 0.0125. Therefore, with the adjusted alpha, we can conclude that at a significance level 0.05 for the experiment, we __do not have enough evidence to reject the null hypothesis__. The Type I error rate is now a more reasonable 0.049.

__2.4. It is is desired to compare mean weights of the 4 methods. This is to be done first by performing 
pairwise comparisons of mean weight for the different methods. What test should be used for these 
comparisons?__

As we discussed before, since we only have 1% of the data, our sample size is relatively small, <br>
We can consider the samples to be independent as they were collected randomly<br>
We are interested to know how the mean weights differ for each of these pairs, so assuming equal variance, we can use a __two sample two tailed equal variance t-test__.

__2.5. Report the p-values from all pairwise comparisons. Include your R code for this question.__

In [97]:

def test(df, pair):
    df_pm1,df_pm2 = [df.loc[(df["Method"] == x)] for x in pair]
    pooled_var = (((len(df_pm1)-1)*df_pm1["Weight"].var())+((len(df_pm2)-1)*df_pm2["Weight"].var()))/(len(df_pm1)+len(df_pm2)-2)
    test_statistic = (df_pm1["Weight"].mean()-df_pm2["Weight"].mean())/((pooled_var*((1/len(df_pm1))+(1/len(df_pm2))))**0.5)
    p_val = stats.t.sf(abs(test_statistic), df=len(df_pm1)+len(df_pm2)-2)*2
    print("For Pair (%s,%s) :\n\tTest Statistic = %f \n\t\033[1mp-value\033[0m = %f"%(pair[0],pair[1],test_statistic,p_val))

for i in range(3):
    for j in range(i+1,4):
        test(df,[processing_methods[i],processing_methods[j]])

For Pair (A,B) :
	Test Statistic = 0.411090 
	p-value = 0.681606
For Pair (A,C) :
	Test Statistic = -1.264579 
	p-value = 0.208185
For Pair (A,D) :
	Test Statistic = -2.086270 
	p-value = 0.039004
For Pair (B,C) :
	Test Statistic = -1.567992 
	p-value = 0.119190
For Pair (B,D) :
	Test Statistic = -2.328241 
	p-value = 0.021506
For Pair (C,D) :
	Test Statistic = -0.884224 
	p-value = 0.378437


__2.6. Apply a Bonferroni correction to your results of the previous question to account for inflation of type I 
error rate due to multiple testing. What conclusion would you draw from these results? Would you reject 
the null hypothesis of no difference between any pair of means among the 4 methods, at significance level 
0.05?__

In [102]:
error = 1 - pow(1-0.05,6)
alpha_bonferroni = 0.05/6 #0.00833
error_new = 1 - pow(1-alpha_bonferroni,6)
print('\033[1m'+"Type I error rate before correction : %f \nType I error rate after correction : %f"%(error,error_new))

Type I error rate before correction : 0.264908 
Type I error rate after correction : 0.048970


In __2.5__, with a significance level of 0.05, we see that for the pairs (A,D) and (B,D) the p-value is less than 0.05, which is why we have to __reject__ the null hypothesis that the mean for all methods is equal. Also, before the correction due to multiple tests, the Type I error rate is 0.264, which is inflated because we associated an alpha of 0.05 to each indidvidual test.

Now, __After Bonferroni Correction__, the new significance level is 0.0083 for each individual test, All the p-values from the possible pairs are greater than 0.0083. Therefore, with the adjusted alpha, we can conclude that at a significance level 0.05 for the experiment, we __do not have enough evidence to reject the null hypothesis__. The Type I error rate is now a more reasonable 0.048.

__2.7. Compare the mean weights for the 4 methods using ANOVA. State the F-statistic and the p-value for the 
F-test. Include your R code for this question.__

In [103]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

m = ols('Weight ~ Method',data=df).fit()
anova = sm.stats.anova_lm(m,typ=1)
print(anova)
print('\033[1m'+"\n\nF statistic : %f \np value : %f"%(anova['F'][0],anova['PR(>F)'][0]))

             df     sum_sq   mean_sq         F    PR(>F)
Method      3.0   1.289595  0.429865  2.616943  0.051475
Residual  261.0  42.872443  0.164262       NaN       NaN


F statistic : 2.616943 
p value : 0.051475


__2.8. What do you conclude from the ANOVA?__

In the slides, ANOVA is defined as : "Analysis of Variance is designed to provide a single test of a null
hypothesis of equal group means with a desired significance level. It is a
generalization of the equal-variance t-test to the case where the number of
means to be compared is greater than 2"

<br>
Here we use the F-test to determine is our group means are equal. The p-value in our experiment 0.05147 is greater than 0.05 (alpha). Therefore, our null hypothesis __cannot be rejected__ because we do not have enough evidence to prove otherwise.

__2.9. How does your conclusion from ANOVA compare to the conclusion from the pairwise comparisons?__

For the pairwaise comparisions without the bonferroni correction, we rejected the null hypothesis but for both ANOVA and two tailed equal variance t-test we concluded that we do not have enough evidence to reject the null hypothesis. Therefore, bonferroni corrected pairwise test yields the same result as ANOVA.